utils/atomic-data-publisher/src/database.rs

```diff
@@ -0,0 +1,797 @@
+
+    let chunk_size = if block_diff <= MIN_CHUNK_SIZE {
+      block_diff
+    } else {
+      // Scale chunk size logarithmically: roughly 10% of remaining blocks, with bounds
+      let scaled_chunk = (block_diff as f64 * DEFAULT_CHUNK_PERCENTAGE) as u64;
+      scaled_chunk.clamp(MIN_CHUNK_SIZE, MAX_CHUNK_SIZE)
+    };
+
+    std::cmp::min(last_processed_block + chunk_size, max_available_block)
+  }
+
+  async fn execute_query(
+    &self,
+    job: &PollingJob,
+    last_processed_block: u64,
+    target_block: u64,
+  ) -> Result<Vec<sqlx::postgres::PgRow>, AtomicDataError> {
+    crate::queries::AtomicHotspotQueries::validate_query_name(&job.query_name)
+      .map_err(|e| AtomicDataError::QueryValidationError(format!("Query validation failed for '{}': {}", job.query_name, e)))?;
+
+    let query = crate::queries::AtomicHotspotQueries::get_query(&job.query_name)
+      .ok_or_else(|| AtomicDataError::QueryValidationError(format!("Query not found for '{}'", job.query_name)))?;
+
+    let query_start = std::time::Instant::now();
+
+    let rows = if job.query_name == "construct_atomic_hotspots" {
+      let hotspot_type = job
+        .parameters
+        .get("hotspot_type")
+        .and_then(|v| v.as_str())
+        .ok_or_else(|| AtomicDataError::InvalidData("hotspot_type parameter required for hotspot queries".to_string()))?;
+
+      match hotspot_type {
+        "iot" | "mobile" => {},
+        _ => return Err(AtomicDataError::InvalidData(format!("Invalid hotspot_type: '{}'. Must be 'iot' or 'mobile'", hotspot_type))),
+      }
+
+      info!(
+        "Querying job '{}' with query '{}' for hotspot_type '{}', processing blocks {} to {} ({} blocks)",
+        job.name, job.query_name, hotspot_type, last_processed_block, target_block, target_block - last_processed_block
+      );
+
+      sqlx::query(query)
+        .bind(hotspot_type)
+        .bind(last_processed_block as i64)
+        .bind(target_block as i64)
+        .fetch_all(&self.pool)
+        .await?
+    } else {
+      info!(
+        "Querying job '{}' with query '{}', processing blocks {} to {} ({} blocks)",
+        job.name, job.query_name, last_processed_block, target_block, target_block - last_processed_block
+      );
+
+      sqlx::query(query)
+        .bind(last_processed_block as i64)
+        .bind(target_block as i64)
+        .fetch_all(&self.pool)
+        .await?
+    };
+
+    let query_duration = query_start.elapsed().as_secs_f64();
+    if let Some(ref metrics) = self.metrics {
+      metrics.observe_database_query_duration(query_duration);
+    }
+
+    Ok(rows)
+  }
+
+  fn process_query_results(&self, rows: &[sqlx::postgres::PgRow], job_name: &str) -> Vec<ChangeRecord> {
+    let mut changes = Vec::with_capacity(rows.len());
+
+    for row in rows {
+      let mut atomic_data_array = Vec::with_capacity(1);
+      atomic_data_array.push(row.get::<serde_json::Value, _>("atomic_data"));
```

Comment:
why does this need to be an array if it's always length one? This function feels like one ChangeRecord should include all rows?